pkg/frr/utils.go

```diff
@@ -0,0 +1,102 @@
+package frr
+
+func Chunk[E any](values []E, size int) [][]E {
+	if size <= 0 {
+		panic("size must be > 0")
+	}
+
+	var chunks [][]E
+	for remaining := len(values); remaining > 0; remaining = len(values) {
+		if remaining < size {
+			size = remaining
+		}
+
+		chunks = append(chunks, values[:size])
+		values = values[size:]
+	}
+
+	return chunks
+}
+
+func DeleteByIndex[T any](slice []T, index int) []T {
+	sliceLastIndex := len(slice) - 1
+
+	if index != sliceLastIndex {
+		slice[index] = slice[sliceLastIndex]
+	}
+
+	return slice[:sliceLastIndex]
+}
+
+func DeleteElementsByIndices[T any](slice []T, indices []int) []T {
+	indicesMap := make(map[int]int)
+
+	for _, index := range indices {
+		indicesMap[index] = index
+	}
+
+	lastIndex := len(slice) - 1
+	backIndex := lastIndex
+
+	for _, index := range indices {
+		if index < 0 || index > lastIndex {
+			continue
+		}
+
+		mappedIndex := indicesMap[index]
+
+		if mappedIndex == -1 {
+			continue
+		}
+
+		if mappedIndex != backIndex {
+			slice[mappedIndex] = slice[backIndex]
+
+			indicesMap[backIndex] = indicesMap[mappedIndex]
+		}
+
+		indicesMap[index] = -1
+
+		backIndex--
+	}
+
+	return slice[:backIndex+1]
+}
+
+func Filter[T any](slice []T, predicate func(T) bool) []T {
+	indices := make([]int, 0, len(slice))
+
+	for index, element := range slice {
+		if !predicate(element) {
+			indices = append(indices, index)
+		}
+	}
+
+	return DeleteElementsByIndices(slice, indices)
+}
```

Comment:
How about just doing something like 
```suggestion
func Filter[T any](slice []T, predicate func(T) bool) (res []T) {
	for _, elem := range slice {
		if predicate(elem) {
			res = append(res, elem)
		}
	}
	return res
}
```
Then we could get rid of the DeleteElementsByIndices function and its complexity (and it's probably faster)